src/mapeo-project.js

```diff
@@ -749,6 +761,389 @@ export class MapeoProject extends TypedEmitter {
+   * @param {boolean} [options.tracks=true] Whether all tracks and their observations should be exported
+   * @param {string} [options.lang]
+   * @returns {Promise<string>} The full path that the file was exported at
+   */
+  async exportGeoJSONFile(
+    exportFolder,
+    { observations = true, tracks = true, lang } = {}
+  ) {
+    const fileName = await this[kGeoJSONFileName](observations, tracks)
+    const filePath = path.join(exportFolder, fileName)
+    const source = this[kExportGeoJSONStream]({ observations, tracks, lang })
+    const sink = createWriteStream(filePath)
+    await pipelinePromise(source, sink)
+
+    return filePath
+  }
+
+  /**
+   * @param {Attachment} attachment
+   * @returns {Promise<null | Readable<Uint8Array|Buffer>>}
+   */
+  async #tryExportBlob(attachment) {
+    // Audio must not have variants
+    for (const variant of VARIANT_EXPORT_ORDER) {
+      const blobId = buildBlobId(attachment, variant)
+      const entry = await this.#blobStore.entry(blobId)
+      if (!entry) continue
+      // Download
+      // @ts-expect-error
+      return this.#blobStore.createReadStream(blobId)
+    }
+
+    return null
+  }
+
+  /**
+   * @param {ZipArchive} archive
+   * @param {Object} [options={}]
+   * @param {boolean} [options.observations=true] Whether observations should be exported
+   * @param {boolean} [options.tracks=true] Whether all tracks and their observations should be exported
+   * @param {boolean} [options.attachments=true] Whether all attachments for observations should be exported
+   * @param {string} [options.lang]
+   * @returns {Promise<void>}
+   */
+  async #exportToArchive(
+    archive,
+    { observations = true, tracks = true, attachments = true, lang } = {}
+  ) {
+    // GeoJSON
+    const geoJSONFileName = await this[kGeoJSONFileName](observations, tracks)
+    const seenAttachments = new Map()
+    const geoJSONStream = this[kExportGeoJSONStream]({
+      observations,
+      tracks,
+      lang,
+      seenAttachments,
+    })
+
+    // @ts-expect-error
+    await archive.entry(geoJSONStream, { name: geoJSONFileName })
+
+    const missingAttachments = []
+    // Attachments
+    if (attachments) {
+      const mediaFolder = this.#exportPrefix('Media') + '/'
+      for (const attachment of seenAttachments.values()) {
+        const stream = await this.#tryExportBlob(attachment)
+        if (stream === null) {
+          missingAttachments.push(attachment)
+          continue
+        }
+
+        // @ts-expect-error
+        await archive.entry(stream, { name: mediaFolder + attachment.name })
+      }
+
+      if (missingAttachments.length) {
+        this.#l.log(`Found missing attachments during export`)
+        const missingContent = missingAttachments
+          .map((attachment) => JSON.stringify(attachment))
+          .join('\n')
+
+        await archive.entry(missingContent, {
+          name: mediaFolder + 'missing.ndjson',
```

Comment:
is the idea that this file could eventually describe other kinds of content aside from attachments? just wondering if a more explicit file name is helpful or not.

Reply:
The spec doesn't get into it too deeply, it's just supposed to contain the missing items.

Originally it was described as `missing.txt` but IMO an ndjson file would be better for machine parsing and rendering missing attachment info. Not obvious how this could mix with existing geojson tools.

https://www.notion.so/digidem/Data-Export-mobile-fe0a747ef76e42c5a3ca32736bd4e82b